File: C++/26.MVC模式/MVCPatternDemo.cpp
/**
 * @file MVCPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-12-02
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "StudentController.h"

static std::shared_ptr<Student> RetrieveStudentFromDatabase() {
    std::shared_ptr<Student> student = std::make_shared<Student>();
    student->SetName("Robert");
    student->SetRollNo("10");
    return student;
}

int main(int argc, char const *argv[]) {
    std::shared_ptr<Student> model = RetrieveStudentFromDatabase();
    std::shared_ptr<StudentView> view = std::make_shared<StudentView>();
    std::shared_ptr<StudentController> controller = std::make_shared<StudentController>(model, view);
    controller->UpdateView();
    controller->SetStudentName("John");
    controller->UpdateView();
    return 0;
}

File: Objective-C/26.MVCPattern/MVCPatternDemo.m
//
//  MVCPatternDemo.m
//  26.MVCPattern
//
//  Created by Layne on 2020/11/21.
//

#import "StudentController.h"

static Student* retrieveStudentFromDatabase() {
    Student* student = [[Student alloc] init];
    student.name = @"Robert";
    student.rollNo = @"10";
    return student;
}

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        //从数据库获取学生记录
        Student* model = retrieveStudentFromDatabase();

        //创建一个视图：把学生详细信息输出到控制台
        StudentView* view = [[StudentView alloc] init];

        StudentController* controller = [[StudentController alloc] initWithModel:model View:view];
        [controller updateView];

        //更新模型数据
        [controller setStudentName:@"John"];
        [controller updateView];
    }
    return 0;
}
